_assets/patches/patcher

```diff
@@ -32,6 +36,54 @@ while getopts :prv opt; do
 		echo "Reverted all."
 		exit
 		;;
+	c)
+		applied=()
+		unapplied=()
+		# Finds applied patches using reverse order and
+		# notes them down.
+		for ((i=${#patches[@]}-1; i>=0; i--)); do
+			f=${patches[$i]}
+			git apply "$f" --directory="$basepath" -R > /dev/null 2>&1
+			if [ $? -eq 1 ]; then
+				unapplied+=("$f")
+			else
+				applied+=("$f")
+			fi
+		done
+		# Applies reverted patches back again.
+		for ((i=${#applied[@]}-1; i>=0; i--)); do
+			f=${applied[$i]}
+			git apply "$f" --directory="$basepath" > /dev/null 2>&1
+		done
+		# Sorts out new patches' paths by comparing with "develop" branch.
+		fromDevelop=($(git diff develop --stat | grep "\\.patch" |
+			while IFS=" " read -r -a line; do
+				path="$(pwd)/${line[0]}"
+				echo "$path"
+			done
+		))
+		# Also does the same with uncommitted.
+		uncommitted=($(git status -u --porcelain | grep "\\.patch" |
+			while IFS=" " read -r -a line; do
+				length=${#line[@]}
+				path="$(pwd)/${line[$((length - 1))]}"
+				echo "$path"
+			done
+		))
+		newPatches=( "${fromDevelop[@]}" "${uncommitted[@]}" )
+		# Checks new patches and exits with 1 if there are unapplied.
```

Comment:
what would happen if we removed a patch?

Reply:
Nothing problematic. Looks like there's a flaw about that - ideally, it should exclude the removed ones - but it doesn't affect the result. Any patch change (addition, removal) is detected as a new patch but it's not counted if it's currently not residing as an unapplied patch in the directory.

Another question is, what happens if there's a non-new & unapplied patch in the directory (since it will get ignored) but that's not a big problem because `make dep-ensure` keeps things on track.